xdsl/dialects/scf.py

```diff
@@ -50,6 +58,78 @@ def get(cond: SSAValue | Operation,
+
+    res: Annotated[VarOpResult, AnyAttr()]
+
+    body: Region
+
+    def verify_(self):
+        if (len(self.iter_args) + 1) != len(self.body.blocks[0].args):
+            raise VerifyException(
+                f"Wrong number of block arguments, expected {len(self.iter_args)+1}, got "
+                f"{len(self.body.blocks[0].args)}. The body must have the induction "
+                f"variable and loop-carried variables as arguments.")
+        if len(self.operands) < 3:
+            raise VerifyException(
+                "Expected at least 3 arguments. A scf.for must have a lower- and "
+                "upper-bound, and a step.")
+        if self.body.blocks[0].args[0].typ != IndexType():
+            raise VerifyException(
+                f"Block argument with wrong type, expected {IndexType()}, "
+                f"got {self.body.blocks[0].args[0].typ}. The first "
+                f"argument of the body must be the induction variable.")
+        for idx, arg in enumerate(self.iter_args):
+            if self.body.blocks[0].args[idx + 1].typ != arg.typ:
+                raise VerifyException(
+                    f"Block arguments with wrong type, expected {arg.typ}, "
+                    f"got {self.body.blocks[0].args[idx].typ}. Arguments after the "
+                    f"induction variable must match the carried variables.")
+        if len(self.iter_args) > 0:
+            if len(self.body.blocks[0].ops) == 0 or not isinstance(
+                    self.body.blocks[0].ops[-1], Yield):
+                raise VerifyException(
+                    "The scf.for's body does not end with a scf.yield. A scf.for loop "
+                    "with loop-carried variables must yield their values at the end of "
+                    "its body.")
+        if len(self.body.blocks[0].ops) > 0 and isinstance(
+                self.body.blocks[0].ops[-1], Yield):
+            yieldop = self.body.blocks[0].ops[-1]
+            if len(yieldop.arguments) != len(self.iter_args):
+                raise VerifyException(
+                    f"Expected {len(self.iter_args)} args, got {len(yieldop.arguments)}. "
+                    f"The scf.for must yield its carried variables.")
+            for idx, arg in enumerate(yieldop.arguments):
+                if self.iter_args[idx].typ != arg.typ:
+                    raise VerifyException(
+                        f"Expected {self.iter_args[idx].typ}, got {arg.typ}. The "
+                        f"scf.for's scf.yield must match carried variables types."
+                    )
+
+    @staticmethod
+    def get(
+        lb: IndexType | Operation,
+        ub: IndexType | Operation,
+        step: IndexType | Operation,
+        iter_args: List[Attribute],
+        body: Region | List[Operation] | List[Block],
+    ) -> For:
+        op = For.build(
+            operands=[lb, ub, step, iter_args],
+            result_types=[list(map(lambda a: a.result.typ, iter_args))],
```

Comment:
This looks a bit concerning, is this a hack, or is this just type system being weird?

Reply:
It's more me being confused, and I think I'll just do otherwise for now.
A requirement of `scf.for` is having the same result type(s) as its iter_args types. So I wanted to just infer the former from the latter in `get`, but I don't see how to do that making the type system happy with the current codebase, and changing that is clearly out of scope of this PR.